### PR TITLE
Implement batch embeddings for `SCVI`

### PR DESCRIPTION
**Pull request by martinkim0**
-   [x] Closes #xxxx (Replace xxxx with the GitHub issue number)
-   [x] Tests added and passed if fixing a bug or adding a new feature
-   [x] All code checks passed.
-   [x] Added type annotations to new arguments/methods/functions.
-   [x] Added an entry in the latest `docs/release_notes/index.md` file if fixing a bug or adding a new feature.
-   [x] If the changes are patches for a version, I have added the `on-merge: backport to x.x.x` label.

### CODE REVIEW

File: scvi/module/_vae.py [hunk 1]
```diff
@@ -166,6 +173,18 @@ def __init__(
...
+        if use_batch_embedding:
+            if n_batch < 2:
+                raise ValueError(
+                    "`adata` must have more than 1 batch when passing in "
+                    "`use_batch_embedding=True`."
+                )
+
+            self.batch_embedding = nn.Embedding(n_batch, batch_embedding_dim)
+            # fclayers expects size of batch input == n_batch
+            n_batch = batch_embedding_dim
```
Comment (canergen) [inline scvi/module/_vae.py:186]: Does the order make sense here?

File: scvi/model/base/_vaemixin.py [hunk 2]
```diff
@@ -210,3 +213,45 @@ def get_latent_representation(
...
+        Array of batch representations of shape ``(n_batches, batch_embedding_dim)``.
+        """
+        self._check_if_trained(warn=False)
+
+        if not hasattr(self.module, "batch_embedding"):
+            raise NotImplementedError("Model does not support batch embeddings.")
+        if self.module.batch_embedding is None:
+            raise ValueError("Model was not trained with batch embeddings.")
+
+        adata = self._validate_anndata(adata)
```
Comment (canergen) [inline scvi/model/base/_vaemixin.py:244]: Add a warning if batch_key has similar name but actually describes different batches the result will be non-sense.

File: scvi/model/base/_vaemixin.py [hunk 3]
```diff
@@ -210,3 +213,45 @@ def get_latent_representation(
...
+        cat_mapping = batch_state_registry.categorical_mapping
+
+        if batch_keys is None:
+            batch_indexes = np.arange(len(cat_mapping))
+        elif not all((key in cat_mapping) for key in batch_keys):
+            raise ValueError("``batch_keys`` contains keys not present in ``adata``.")
+        else:
+            batch_indexes = np.where(cat_mapping == batch_keys)[0]
+
+        batch_embeddings = self.module.batch_embedding.weight.detach().cpu().numpy()
```
Comment (canergen) [inline scvi/model/base/_vaemixin.py:256]: Potentially leave this on GPU. Run embedding and retrieve the embeddings. This object can be rather large if you have a large number of batches.

File: scvi/module/_vae.py (hunk 1 above)
Comment (canergen) [inline scvi/module/_vae.py:186]: Currently, categorical_covariates are not batch_embedded. I guess it's fine but depending on the use case it might be useful.

File: scvi/module/_vae.py [hunk 4]
```diff
@@ -279,6 +302,10 @@ def _get_generative_input(self, tensors, inference_outputs):
             else None
         )
 
+        if self.batch_embedding is not None:
```
Comment (canergen) [inline scvi/module/_vae.py:305]: Out of curiosity: Is the embedding optimized to be good for the encoder or the decoder?

**Review by canergen — COMMENTED** [review]
We should add a tutorial. Code changes LGTM. Have you done some benchmarking of it?

File: scvi/model/base/_vaemixin.py (hunk 2 above)
Comment (martinkim0) [reply scvi/model/base/_vaemixin.py:244]: Thanks for taking a look! I'm not following, could you clarify what you mean?

File: scvi/model/base/_vaemixin.py (hunk 3 above)
Comment (martinkim0) [reply scvi/model/base/_vaemixin.py]: Yeah it could potentially be large depending on the number of batches. Two reasons why I support moving it to CPU: (1) it would be more consistent with `get_latent_representation`, which returns numpy arrays, (2) and by default the batch embeddings are guaranteed to be smaller than the latent representation since `batch_embedding_dim=5`.

File: scvi/model/base/_vaemixin.py (hunk 2 above)
Comment (canergen) [reply scvi/model/base/_vaemixin.py:244]: If a user provides a new adata object with different batches, in which batches are e.g. just numbered, it will output batch embeddings for those that have no meaning. I think it's important that we allow using a new adata object but we could issue a warning.

File: scvi/module/_vae.py (hunk 1 above)
Comment (martinkim0) [reply scvi/module/_vae.py:186]: Good point! Do you think supporting other categorical covariates would be an interesting use case? In terms of the implementation, it would make sense if we supported embedding other covariates, we would just have to think a bit more to make the API non-redundant.

File: scvi/model/base/_vaemixin.py (hunk 3 above)
Comment (canergen) [reply scvi/model/base/_vaemixin.py]: We are not copying a 5-dimensional batch_embedding but a n_batchxbatch_embedding_dim dimensional tensor here. However, copying batch_indices to GPU will not reduce the amount that needs to be transferred.

File: scvi/model/base/_vaemixin.py (hunk 2 above)
Comment (canergen) [reply scvi/model/base/_vaemixin.py:244]: I guess the same thing would happen without batch_embeddings as well. Marking it as resolved.

**Review by canergen — APPROVED** [review]
We should add a tutorial and see that the results are fine there. Code LGTM.

File: scvi/model/base/_vaemixin.py (hunk 2 above)
Comment (martinkim0) [reply scvi/model/base/_vaemixin.py:244]: Ah I see what you mean now. I have to think more about this.

File: scvi/module/_vae.py (hunk 4 above)
Comment (martinkim0) [reply scvi/module/_vae.py:305]: The embeddings are by default fed into the decoder only

File: scvi/module/_vae.py (hunk 1 above)
Comment (canergen) [reply scvi/module/_vae.py:186]: I think batch_key and categorical_covariate in the same low-dimensional batch_embedding are fine (simply adding up both embeddings). It would be good to keep the names separate though. I have to think more about the use case. It requires cases with a large number of categorical covariate values and I haven't used those cases.

File: scvi/model/base/_vaemixin.py (hunk 2 above)
Comment (martinkim0) [reply scvi/model/base/_vaemixin.py:244]: I feel like the simplest solution to this would be removing the `adata` argument in `get_batch_representation`.
